climada_petals/entity/exposures/openstreetmap/osm_dataloader.py

```diff
@@ -0,0 +1,945 @@
+
+        return query
+
+    def retrieve(self, geo_type, constraint_dict):
+        """
+        Function to extract geometries and tag info for entires in the OSM file
+        matching certain OSM key-value constraints.
+        from an OpenStreetMap osm.pbf file.
+        adapted from BenDickens/trails repo
+        (https://github.com/BenDickens/trails.git, see extract.py)
+        
+
+        Parameters
+        ----------
+        geo_type : str
+            Type of geometry to retrieve. One of [points, lines, multipolygons]
+        constraint_dict :  dict
+            A dict with the keys "osm_keys" and "osm_query". osm_keys contains
+            a list with all the osm keys that should be reported as columns in
+            the output gdf. 
+            osm_query contains an osm query string of the syntax
+            "key(='value') (and/or further queries)".
+            See examples in DICT_CIS_OSM in case of doubt.
+
+        Returns
+        -------
+        gpd.GeoDataFrame
+            A gdf with all results from the osm.pbf file matching the 
+            specified constraints.
+
+        Note
+        ----
+        1) The keys that are searchable are specified in the osmconf.ini file.
+        Make sure that they exist in the attributes=... paragraph under the 
+        respective geometry section. 
+        For example, to retrieve multipolygons with building='yes', 
+        building must be in the attributes under
+        the [multipolygons] section of the file. You can find it in the same 
+        folder as the osm_dataloader.py module is located.
+        2) OSM keys that have : in their name must be changed to _ in the 
+                search dict, but not in the osmconf.ini
+        E.g. tower:type is called tower_type, since it would interfere with the
+        SQL syntax otherwise, but still tower:type in the osmconf.ini
+        
+        See also
+        --------
+        https://taginfo.openstreetmap.org/ to check what keys and key/value
+        pairs are valid.
+        https://overpass-turbo.eu/ for a direct visual output of the query,
+        and to quickly check the validity. The wizard can help you find the
+        correct keys / values you are looking for.
+        """
+        driver = ogr.GetDriverByName('OSM')
+        data = driver.Open(self.osm_path)
+        query = self._query_builder(geo_type, constraint_dict)
+        sql_lyr = data.ExecuteSQL(query)
+        features = []
+        if data is not None:
+            LOGGER.info('query is finished, lets start the loop')
+            for feature in tqdm(sql_lyr, desc=f'extract {geo_type}'):
+                try:
+                    fields = []
+                    for key in ['osm_id', *constraint_dict['osm_keys']]:
+                        fields.append(feature.GetField(key))
+                    geom = shapely.wkb.loads(feature.geometry().ExportToWkb())
+                    if geom is None:
+                        continue
```

Comment:
what are the situations where geom will be none? is this common?
it feels weird to throw a warning on line 658 when you skip a feature but not here

Reply:
yeah there are actually instances where there is no geometry reported. I guess for some use cases it could still be useful that a feature is "there" somewhere.. not sure^^

Reply:
i have another question about this: in the tutorial the geodataframe for zurich buildings includes a row with the geometry GEOMETRYCOLLECTION EMPTY
would this count as `geom is None`? if so we should be dropping the nones in all methods or none

Reply:
right. For the cases where it has a GEOMETRYCOLLECTION EMPTY, one can check with geometry.empty (not `None`, however). I'm doing this in the lines_polys exposuremaker functions (checking, and skipping those for now)

Reply:
if the exposuremaker functions handle this, maybe we just keep the empty geometries here. could be the easiest solution.